apollo-federation/src/sources/connect/validation/entity.rs

```diff
@@ -15,125 +13,141 @@ use apollo_compiler::Node;
 
-    if let Some(entity_arg) = connect_directive
+    let Some(entity_arg) = connect_directive
         .arguments
         .iter()
         .find(|arg| arg.name == CONNECT_ENTITY_ARGUMENT_NAME)
+    else {
+        return Ok(());
+    };
+    let entity_arg_value = &entity_arg.value;
+    if !entity_arg_value
+        .to_bool()
+        .is_some_and(|entity_arg_value| entity_arg_value)
     {
-        let entity_arg_value = &entity_arg.value;
-        if entity_arg_value
-            .to_bool()
-            .is_some_and(|entity_arg_value| entity_arg_value)
-        {
-            if category != ObjectCategory::Query {
-                messages.push(Message {
-                    code: Code::EntityNotOnRootQuery,
-                    message: format!(
-                        "{coordinate} is invalid. Entity resolvers can only be declared on root `Query` fields.",
-                        coordinate = connect_directive_entity_argument_coordinate(connect_directive_name, entity_arg_value.as_ref(), object, &field.name)
-                    ),
-                    locations: entity_arg.line_column_range(&schema.sources)
-                        .into_iter()
-                        .collect(),
-                })
-                // TODO: Allow interfaces
-            } else if field.ty.is_list()
-                || schema.get_object(field.ty.inner_named_type()).is_none()
-                || field.ty.is_non_null()
-            {
-                messages.push(Message {
-                    code: Code::EntityTypeInvalid,
-                    message: format!(
-                        "{coordinate} is invalid. Entity connectors must return non-list, nullable, object types.",
-                        coordinate = connect_directive_entity_argument_coordinate(
-                            connect_directive_name,
-                            entity_arg_value.as_ref(),
-                            object,
-                            &field.name
-                        )
-                    ),
-                    locations: entity_arg
-                        .line_column_range(&schema.sources)
-                        .into_iter()
-                        .collect(),
-                })
+        // This is not an entity resolver
+        return Ok(());
+    }
+
+    if category != ObjectCategory::Query {
+        return Err(
+            Message {
+                code: Code::EntityNotOnRootQuery,
+                message: format!(
+                    "{coordinate} is invalid. Entity resolvers can only be declared on root `Query` fields.",
+                    coordinate = connect_directive_entity_argument_coordinate(connect_directive_name, entity_arg_value.as_ref(), object, &field.name)
+                ),
+                locations: entity_arg.line_column_range(&schema.sources)
+                    .into_iter()
+                    .collect(),
             }
+        );
+        // TODO: Allow interfaces
```

Comment:
is this comment misplaced?

Reply:
The comment was there originally, I just left it. TBH I'm not sure what interfaces have to do with being on the root query type, though...

```suggestion
```